File: day2/day2.ts
import { getExampleInput, getPuzzleInput } from "../helper/inputs";

const example: string[] = getExampleInput(__dirname);
const input: string[] = getPuzzleInput(__dirname);

type Set = {
  red: number;
  blue: number;
  green: number;
};

class Game {
  id: number;
  sets: Set[];

  constructor(id: number) {
    this.id = id;
    this.sets = [];
  }

  getMaxRed(): number {
    return this.sets.reduce(function (p, c) {
      return p && p.red > c.red ? p : c;
    }).red;
  }
  getMaxBlue(): number {
    return this.sets.reduce(function (p, c) {
      return p && p.blue > c.blue ? p : c;
    }).blue;
  }
  getMaxGreen(): number {
    return this.sets.reduce(function (p, c) {
      return p && p.green > c.green ? p : c;
    }).green;
  }
}

console.log(`Part 1 example: ${solve1(example)}`);
console.log(`Part 1 solution: ${solve1(input)}`);
console.log(`Part 2 example: ${solve2(example)}`);
console.log(`Part 2 solution: ${solve2(input)}`);

function solve1(lines: string[], total = 0) {
  let games: Game[] = parseGames(lines);

  games = games.filter(
    (game) =>
      game.getMaxRed() < 13 && game.getMaxGreen() < 14 && game.getMaxBlue() < 15
  );
  games.forEach((g) => (total = total + g.id));
  return total;
}

function solve2(lines: string[], total = 0) {
  let games: Game[] = parseGames(lines);
  games.forEach(
    (g) => (total = total + g.getMaxBlue() * g.getMaxRed() * g.getMaxGreen())
  );
  return total;
}

function parseGames(lines: string[]) {
  let games: Game[] = [];
  lines.forEach((line) => {
    const arr = line.split(":");
    const gameId = +arr[0].replace("Game ", "");
    let game = new Game(gameId);

    const setsraw = arr[1].split(";");
    setsraw.forEach((s) => {
      s = s.trim();
      game.sets.push(createSet(s));
    });
    games.push(game);
  });
  return games;
}

function createSet(rawset: string) {
  const gemsraw = rawset.split(",");
  let set: Set = { red: 0, blue: 0, green: 0 };

  gemsraw.forEach((gem) => {
    gem = gem.trim();
    const amount = +gem.split(" ")[0];
    const colour = gem.split(" ")[1];

    switch (colour) {
      case "red":
        set.red = amount;
        break;
      case "blue":
        set.blue = amount;
        break;
      case "green":
        set.green = amount;
        break;
    }
  });
  return set;
}

File: day2/day2.jl
include(pwd() * "\\helper\\inputs.jl")
example = getExample(@__DIR__)
input = getInput(@__DIR__)

struct CubeSet
	red::Integer
	green::Integer
	blue::Integer
end

function parseToCubeSet(raw)
	red = green = blue = 0
	for cube in raw
		(amount, colour) = split(lstrip(cube), " ")
		if colour == "red"
			red = parse(Int, amount)
		elseif colour == "green"
			green = parse(Int, amount)
		elseif colour == "blue"
			blue = parse(Int, amount)
		end
	end
	CubeSet(red, green, blue)
end

function solve(lines, part)
	lines = split.(lines, ":")
	games = []
	sum = 0
	for line in lines
		line[begin] = replace(line[begin], "Game " => "")
		setsraw = split(line[end], ";")
		cubeSet = parseToCubeSet.(split.(setsraw, ","))
		push!(games, (id = line[begin], set = cubeSet))
	end
	for game in games
		maxred = maximum(getfield.(game.set, :red))
		maxgreen = maximum(getfield.(game.set, :green))
		maxblue = maximum(getfield.(game.set, :blue))
		if part == 1 && maxred < 13 && maxgreen < 14 && maxblue < 15
			sum += parse(Int, game.id)
		end
		if part == 2
			sum += (maxred * maxgreen * maxblue)
		end
	end
	return sum
end

println("Part 1 example : $(solve(example, 1))")
println("Part 1 solution: $(solve(input,1))")
println("Part 2 example : $(solve(example,2))")
println("Part 2 solution: $(solve(input,2))")

